cmd/utils/jeth.go

```diff
@@ -129,76 +110,62 @@ func (self *Jeth) NewAccount(call otto.FunctionCall) (response otto.Value) {
 }
 
 func (self *Jeth) Send(call otto.FunctionCall) (response otto.Value) {
-	reqif, err := call.Argument(0).Export()
-	if err != nil {
-		return self.err(call, -32700, err.Error(), nil)
+	ro := call.Argument(0).Object()
+	if ro == nil || (ro.Class() != "Array" && ro.Class() != "Object") {
+		throwJSExeception("Internal Error: request must be an object or array")
 	}
-
-	jsonreq, err := json.Marshal(reqif)
-	var reqs []rpc.JSONRequest
-	batch := true
-	err = json.Unmarshal(jsonreq, &reqs)
+	reqv, err := call.Otto.Call("JSON.stringify", otto.NullValue(), ro)
 	if err != nil {
-		reqs = make([]rpc.JSONRequest, 1)
-		err = json.Unmarshal(jsonreq, &reqs[0])
-		batch = false
+		throwJSExeception(err.Error())
 	}
-
-	call.Otto.Set("response_len", len(reqs))
-	call.Otto.Run("var ret_response = new Array(response_len);")
-
-	for i, req := range reqs {
-		err := self.client.Send(&req)
-		if err != nil {
-			return self.err(call, -32603, err.Error(), req.Id)
-		}
-
-		result := make(map[string]interface{})
-		err = self.client.Recv(&result)
-		if err != nil {
-			return self.err(call, -32603, err.Error(), req.Id)
-		}
-
-		_, isSuccessResponse := result["result"]
-		_, isErrorResponse := result["error"]
-		if !isSuccessResponse && !isErrorResponse {
-			return self.err(call, -32603, fmt.Sprintf("Invalid response"), new(int64))
-		}
-
-		id, _ := result["id"]
-		call.Otto.Set("ret_id", id)
-
-		jsonver, _ := result["jsonrpc"]
-		call.Otto.Set("ret_jsonrpc", jsonver)
-
-		var payload []byte
-		if isSuccessResponse {
-			payload, _ = json.Marshal(result["result"])
-		} else if isErrorResponse {
-			payload, _ = json.Marshal(result["error"])
-		}
-		call.Otto.Set("ret_result", string(payload))
-		call.Otto.Set("response_idx", i)
-
-		response, err = call.Otto.Run(`
-		ret_response[response_idx] = { jsonrpc: ret_jsonrpc, id: ret_id, result: JSON.parse(ret_result) };
-		`)
+	req := json.RawMessage(reqv.String())
+
+	if cb := call.Argument(1); cb.IsFunction() {
+		// Asynchronous call: do the the communication on another goroutine, then switch
+		// back to the event loop and call cb with the response or error.
+		go self.execRequests(&req, func(resp json.RawMessage, err error) {
```

Comment:
The RPC layer is able to execute requests in parallel and can send responses back out of order. It is up to the caller to correlate a response with the requests and call the appropriate callback. This mechanism doesn't match the response id with the request id and can call the wrong callback for the response.

Example:
```
> eth.getAccounts(function(e, accs) { console.log(accs) }); eth.blockNumber
1042569
> 0x079bd9292ae3f362de496bb706994bb13f0f66be,0x66f183060253cfbe45beff1e6e7ebbe318c81e56,0x4ba70599fa368eb9b0720ff6af21315f630a13f1


> eth.getAccounts(function(e, accs) { console.log(accs) }); eth.blockNumber
0xfe889
new BigNumber() not a base 16 number: 
    at L (bignumber.js:3:2876)
    at bignumber.js:3:8435
    at a (bignumber.js:3:389)
    at bignumber.js:3:8425
    at a (bignumber.js:3:796)
    at a (bignumber.js:3:228)
    at web3.js:2236:16
    at web3.js:2097:12
    at web3.js:5701:54
    at get (web3.js:5751:16)
    at <anonymous>:1:1=
```